crates/oxc_linter/src/rules/react/no_danger_with_children.rs

```diff
@@ -0,0 +1,351 @@
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+fn no_danger_with_children_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Only set one of `children` or `props.dangerouslySetInnerHTML`")
+			.with_help("`dangerouslySetInnerHTML` is not compatible with also passing children and React will throw a warning at runtime.")
+			.with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoDangerWithChildren;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallows when a DOM element is using both `children` and `dangerouslySetInnerHTML` properties.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// React will throw a warning if this rule is ignored and both `children` and `dangerouslySetInnerHTML` are used.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```jsx
+    /// <div dangerouslySetInnerHTML={{ __html: "HTML" }}>Children</div>
+    /// React.createElement(
+    ///     "div",
+    ///     { dangerouslySetInnerHTML: { __html: "HTML" } },
+    ///     "Children"
+    /// );
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```jsx
+    /// <div>Children</div>
+    /// <div dangerouslySetInnerHTML={{ __html: "HTML" }} />
+    /// ```
+    NoDangerWithChildren,
+    correctness
+);
+
+impl Rule for NoDangerWithChildren {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.kind() {
+            AstKind::JSXElement(jsx) => {
+                // Either children are passed in as a prop like `children={}` or they are nested between the tags.
+                let has_children = has_jsx_prop(ctx, node, "children")
+                    || (!jsx.children.is_empty() && !is_line_break(&jsx.children[0]));
+                if !has_children {
+                    return;
+                }
+
+                let has_danger_prop = has_jsx_prop(ctx, node, "dangerouslySetInnerHTML");
+                if has_danger_prop {
+                    ctx.diagnostic(no_danger_with_children_diagnostic(jsx.span));
+                }
+            }
+            AstKind::CallExpression(call_expr) => {
+                // Calls with zero or one arguments are safe since they are not proper createElement calls.
+                if call_expr.arguments.len() <= 1 {
+                    return;
+                }
+                let Expression::StaticMemberExpression(callee) = &call_expr.callee else {
+                    return;
+                };
+
+                // Only accept calls like `.createElement(...)`
+                if callee.property.name != "createElement" {
+                    return;
+                }
+
+                let Some(props) = call_expr
+                    .arguments
+                    .get(1)
+                    .expect("Could not get props argument for createElement call")
```

Comment:
Don't panic here. This could reasonably happen in real code; it's best to just return.

Consider using this instead:
```rust
call_expr
  .arguments
  .get(1)
  .and_then(Argument::as_expression)
```

Reply:
Ah good call 👍 Added in https://github.com/oxc-project/oxc/pull/5420/commits/9cd32f688a245a9f89a1e3d7234794c90aba23c0